Support isbits Union array elements and struct fields

quinnj opened the pull request:
Note this is based on the `jn/union-bits-layout` branch.

This is an initial attempt to fix some of the issues reported [here](https://github.com/JuliaLang/julia/issues/22097).

This successfully compiles, but I'm still feeling my way around `/src` code and figuring out if my code paths are even getting hit. The idea here is that when we allocate a `Union{T, S}` array where `T` and `S` are both isbits, we compute the "element size" for the array using `jl_union_isbits`, allocate the buffer for that, then add on an extra "selector" byte for each element after the data buffer. The "selector" bytes indicate which Union element is actually stored in a data buffer element.

If anyone can provide more pointers to what else needs to be updated, or simple ways to ensure I'm actually hitting codepaths, I'd appreciate it. I feel like a bit of a bull in a china shop here 😄 .

Closes https://github.com/JuliaLang/julia/issues/17073

vtjnash committed:
simple pass at giving union fields an optimized layout

unlike codegen, only bitstypes (!isptr) fields are permitted in the union
and the offset count starts from 0 instead of 1
but otherwise the tindex counter is compatible

timholy commented:
Could use a review from @vtjnash.

quinnj commented:
Unfortunately for him, I commandeered his day yesterday to help me w/ the latest commit I pushed 😄

quinnj commented:
Ok, an update here:

* Most of the mutating operations on Vectors should be good to go; that code is a bit hairy, so I tried to hit all the code paths in the new core.jl tests
* I'm seeing a segfault when doing `v = Vector{Union{Int8, Void}}(10); v[1] = 10`, I believe it's due to trying to setindex! w/ a Int64 instead of Int8, and that the changes in codegen aren't doing something right here? @vtjnash do we need to insert a `convert(T, x)::T` like the definition in Base has?
* Still need to update the serialize/deserialize code to be aware of Union isbits arrays
* Still want to add tests for @vtjnash's Union isbits struct field code
* I plan on adding more comments/docs around what we're doing here; it's not too crazy, but I am slightly worried about people who may start using Arrays in non-standard ways and the surprises they might run into; that of course assumes they're actually using Union isbits arrays and doing something not very standard (calling array.c functions for example)
* Benchmarks: not sure if we need to add any to BaseBenchmarks, but I want to do some comparisons vs. DataArrays and NullableArrays just to know what that story looks like
* Discovered that `Array{Union{}, 2}(0)` throws a weird error, I'll open a separate issue about that
* I also meant to ask @vtjnash or @JeffBezanson how Unions' types get sorted, i.e. `Union{Int, Void}` vs. `Union{Void, Int}`. Currently it always get sorted as `Union{Int, Void}`, but I'm wondering if that could be changed (easily?) to sort datatype singleton's first? That would have the nice properties of 1) getting an array of nulls by default in the case of `Vector{Union{T, Null}}`, no filling required and 2) match the bitmask implementations of pandas & apache arrow of using `0 = null` in their bitmasks that accompany their column data.

Otherwise, I do think this code is at a place to be reviewed by @vtjnash & @yuyichao at least in case there's something egregious going on.

tkelman commented on src/datatype.c:
```diff
@@ -216,6 +216,44 @@ unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t)
+
+JL_DLLEXPORT int jl_islayout_inline(jl_value_t *eltype, size_t *fsz, size_t *al)
+ {
+     unsigned countbits = union_isbits(eltype, fsz, al);
```
indent's off here

vchuravy commented on src/datatype.c:
```diff
@@ -216,6 +216,44 @@ unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t)
+    }
+    if (jl_layout_isbits(ty)) {
+        size_t sz = jl_datatype_size(ty);
+        size_t al = ((jl_datatype_t*)ty)->layout->alignment;
```
You can use `jl_datatype_align` for this

vchuravy commented on src/datatype.c:
```diff
@@ -216,6 +216,44 @@ unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t)
+        size_t al = ((jl_datatype_t*)ty)->layout->alignment;
+        if (*nbytes < sz)
+            *nbytes = sz;
+        if (*align < al)
```
IIUC in a bitsunion array we can store elements that are either`A` or `B`?

I would think that the alignment needs to be a multiple of the alignment of `A` and `B`.
As an example, if `A` has alignment 8 and `B` as alignment `12` the alignment would need to be `24` (Joint alignment needs to be divisible by the individual alignments.)

vchuravy replied:
As it has been pointed out to me alignment 12 shouldn't exist (and me encountering that is probably a bug). Nevertheless, we should make sure that the new alignment is compatible with both.

vtjnash replied:
Right, alignments must be powers of N (usually 2), otherwise N * alignment wouldn't be aligned to the alignment. Hence picking the larger alignment is always sufficient.

KristofferC commented:
Added a "Closes https://github.com/JuliaLang/julia/issues/17073" to the top comment.

quinnj commented:
Alrighty, another round of updates here. The new commits address the comments that were posted, general code cleanup, code comments and documentation, and getting tests passing (all Base tests pass for me locally). Apart from ensuring CI is happy, the remaining known issues that I'm aware of are two bugs that I suspect may be related to still-needed inference changes?

Case 1:
```julia
v = Vector{Union{Int8, Void}}(10)
v[1] = 10
```
this currently segfaults on an assert on the TIndex. I don't quite understand how we're getting past the call to convert in the Base.setindex! method
```julia
setindex!(A::Array{T}, x, i1::Int) where {T} = arrayset(A, convert(T,x)::T, i1)
```
Julia 0.6 appropriately fails with the error of `"can't convert argument to Union{Int8, Void}"`.
My efforts to debug this haven't been very fruitful, as it seems once we're in codegen, we're already past the point where the convert should have happened? I'd really appreciate any insights here or help in understanding the interaction w/ codegen.

Case 2:
Found this while writing tests for @vtjnash's work on isbits Union struct field types.
```julia
julia> mutable struct UnionField
           x::Union{Int8, Void}
       end

julia> x = UnionField(nothing)
UnionField(nothing)

julia> x.x = Int8(1)
1

julia> x
UnionField(0)
```
Again, my efforts of trying to step through codegen were pretty futile and I had a hard time figuring out why exactly things weren't getting set correctly. I'll probably try to dig into this error again, but any tips again would be most appreciated.

quinnj committed:
Rebase

quinnj committed:
A few fixes for union isbits struct fields

quinnj committed:
Inline support for isbits union array element types

quinnj committed:
Add tests for isbits union optimizations

quinnj commented on src/cgutils.cpp:
```diff
@@ -2229,11 +2274,26 @@ static void emit_setfield(jl_codectx_t &ctx,
+                Value *ptindex = ctx.builder.CreateGEP(T_int8, emit_bitcast(ctx, addr, T_pint8), ConstantInt::get(T_size, fsz - 1));
+                ctx.builder.CreateStore(tindex, ptindex);
+                // copy data
+                if (!rhs.isghost) {
```
Not sure if this is the most idiomatic way to check for a singleton type.

quinnj commented on src/codegen.cpp:
```diff
@@ -2688,12 +2707,37 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
+                        if (jl_is_uniontype(ety)) {
+                            Value *nbytes = ConstantInt::get(T_size, elsz);
+                            Value *data = emit_arrayptr(ctx, ary, args[1]);
+                            if (!(rhs.typ == jl_bottom_type)) {
```
Not sure if this is the right kind of check here; the issue is when you have something like
```julia
v = Vector{Union{Int8, Void}}(10);
v[1] = Int64(10)
```
this should result in a `convert` error, but it passes thru codegen with `rhs` being `jl_bottom_type`, which ends up throwing an error in `compute_tindex_unboxed` (since Bottom type doesn't match one of the Union types).

quinnj commented:
Alright, this now passes all tests locally and I've fixed all known bugs that I'm aware of. Let's see what CI says.

quinnj commented:
Progress! 2 green checks on travis (linux & osx 64-bit), and just pushed a fix for 32-bit. I'm not sure what happened on appveyor 64-bit, has anyone seen that error before?

tkelman commented:
gc issue, missing root maybe?
```
From worker 9:	WARNING: Method definition Allocations: 152865371 (Pool: 152848789; Big: 16582); GC: 390
	From worker 9:	<?#0000000008B10900::0000000000000000>
	From worker 9:	0000000027A25040: Queued root: 0000000005170C70 :: 0000000008B10DF0 (bits: 3)
	From worker 9:	        of type Array{Any, 1}

... snip

From worker 9:	This application has requested the Runtime to terminate it in an unusual way.
	From worker 9:	Please contact the application's support team for more information.
	From worker 9:	
	From worker 9:	signal (22): SIGABRT
	From worker 9:	while loading C:\projects\julia\julia-2d15e1e9d0\share\julia\test\core.jl, in expression starting on line 5130
	From worker 9:	crt_sig_handler at C:/projects/julia/src\signals-win.c:92
	From worker 9:	raise at C:\windows\system32\msvcrt.dll (unknown line)
	From worker 9:	abort at C:\windows\system32\msvcrt.dll (unknown line)
	From worker 9:	gc_assert_datatype_fail at C:/projects/julia/src\gc.c:1371
	From worker 9:	gc_mark_loop at C:/projects/julia/src\gc.c:2176
	From worker 9:	_jl_gc_collect at C:/projects/julia/src\gc.c:2457
	From worker 9:	jl_gc_collect at C:/projects/julia/src\gc.c:2622
	From worker 9:	jl_gc_pool_alloc at C:/projects/julia/src\gc.c:954 [inlined]
	From worker 9:	jl_gc_alloc_ at C:/projects/julia/src\julia_internal.h:265 [inlined]
	From worker 9:	jl_gc_alloc at C:/projects/julia/src\gc.c:2657
	From worker 9:	jl_new_struct at C:/projects/julia/src\datatype.c:711
	From worker 9:	var_gt at C:/projects/julia/src\subtype.c:498
	From worker 9:	forall_exists_equal at C:/projects/julia/src\subtype.c:938
	From worker 9:	subtype at C:/projects/julia/src\subtype.c:917
```

quinnj commented:
Removed the "WIP" from the PR title and updated to reflect struct field type support. There have been intermittent travis/appveyor failures that are fairly cryptic, some seem unrelated to changes here while others do seem related. I'll keep trying to track down failures, but at this point, those w/ more knowledge might see obvious issues by just reviewing the code.

quinnj commented:
Uh....what's w/ the windows failure? Did I change to a unsupported printf format code for windows or something?

vtjnash commented:
`zu bytes`, eh? Yeah, that's not supported. Just cast to an `int` (or use an `int` for the alignment computation).

quinnj commented:
Has anyone seen the codegen test failure before? https://travis-ci.org/JuliaLang/julia/jobs/254076809

quinnj commented:
Alright, rebased after some recent codegen/ccall changes. Let's see what CI says this time around.

quinnj commented:
Same CI results as #22781, 64-bit linux actually passed all tests then timed out, 32-bit just ran out of time (though got a bit further thru all the tests).

tkelman commented on base/array.jl:
```diff
@@ -109,6 +137,19 @@ function isassigned(a::Array, i::Int...)
+
+For an Array with `isbits` Union elements, return the "selector bytes" that indicate the index
+of the type for each array element, i.e. the type of `A[1]` is `Base.uniontypes(eltype(A))[Base.selectorbytes(A)[1] + 1]`.
+**NOTE**: The actual array selector bytes are returned, meaning if individual elements are modified, the original array will reflect those changes.
```
wrap at 92 chars

StefanKarpinski commented:
I restarted those jobs – since this touches core language code, it's best to be safe.

quinnj commented:
Hmmm, the freebsd error looks legit, I'll put the goggles on and try to figure out what causes a bus error.

iblis17 commented:
Both of bsd workers (ionic[1] and gaebolg[2]) got the bus error...

[1] https://julia.iblis.cnmc.tw/#/builders/1/builds/856
[2] https://julia.iblis.cnmc.tw/#/builders/1/builds/860

vtjnash commented on src/dump.c:
```diff
@@ -570,7 +570,8 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
         if (!ar->flags.ptrarray) {
-            size_t tot = jl_array_len(ar) * ar->elsize;
+            size_t extra = jl_is_uniontype(jl_tparam0(jl_typeof(ar))) ? jl_array_len(ar) : 0;
+            size_t tot = jl_array_len(ar) * ar->elsize + extra;
```
staticdata.c will need this same change

quinnj replied:
I'll take a look.

quinnj replied:
Updated

vtjnash commented on src/gc.c:
```diff
@@ -857,6 +857,8 @@ static size_t array_nbytes(jl_array_t *a)
     else
         sz = a->elsize * jl_array_len(a);
+    if (!a->flags.ptrarray && jl_is_uniontype(jl_tparam0(jl_typeof(a))))
+        sz += jl_array_len(a);
```
add a comment that this accounts for the union bits

vtjnash commented on base/array.jl:
```diff
@@ -109,6 +137,19 @@ function isassigned(a::Array, i::Int...)
+"""
+function selectorbytes(a::Array{T, N}) where {T, N}
+    isbitsunion(T) || return UInt8[]
+    return unsafe_wrap(Array{UInt8, N}, convert(Ptr{UInt8}, pointer(a)) + length(a) * elsize(a), size(a))
```
Actually, since it doesn't appear we use this anywhere, I would prefer not to provide it. It violates our aliasing and gc assumptions. It must not be getting tested fully either right now, since the bytes aren't at `length(a) * elsize(a)` (we don't expose any way to get at the actual location currently – we would need to export something new from C to return it)

quinnj replied:
I see what you mean. It's not used anywhere in Base, but it's something I planned on using in the Feather.jl package at least; the feather file-format stores arrays in contiguous chunks, with separate chunks for their "null-fields", essentially a bitmask, where each bit is set indicating whether an element is null or not. I would just want some way to "set the bits" when deserializing a feather file.

vtjnash replied:
Yes, we'll probably need to provide some way to set these, but exposing this as an array just breaks to many assumptions and optimizations

quinnj replied:
ok, fair. What would be better? A custom `UnionTagArray` type? That just holds a pointer to the start of the selector bytes?

vtjnash replied:
Probably just some sort of special-case constructor that copies over all of the data needed. For reading it out again, I think `typeof(a[i])` may be already optimal.

Exposing a way to convert the selector bytes to `Ptr{UInt8}` would likely be useful however (for serialization routines internals to Julia)

quinnj replied:
I just took it out for now since it's not immediately needed.

vtjnash commented on src/array.c:
```diff
@@ -180,7 +182,7 @@ JL_DLLEXPORT jl_array_t *jl_reshape_array(jl_value_t *atype, jl_array_t *data,
 
     int ndimwords = jl_array_ndimwords(ndims);
-    int tsz = JL_ARRAY_ALIGN(sizeof(jl_array_t) + ndimwords*sizeof(size_t) + sizeof(void*), JL_SMALL_BYTE_ALIGNMENT);
+    int tsz = JL_ARRAY_ALIGN(sizeof(jl_array_t) + ndimwords * sizeof(size_t) + sizeof(void*), JL_SMALL_BYTE_ALIGNMENT);
```
Need to update the setting of `a->maxsize` below to point at the union selector bits

vtjnash commented on src/array.c:
```diff
@@ -485,8 +499,15 @@ JL_DLLEXPORT jl_value_t *jl_arrayref(jl_array_t *a, size_t i)
+        jl_value_t *eltype = (jl_value_t*)jl_tparam0(jl_typeof(a));
+        if (jl_is_uniontype(eltype)) {
+            // isbits union selector bytes are always stored directly after the last array element
+            uint8_t sel = ((uint8_t*)a->data)[jl_array_len(a) * a->elsize + i];
```
I thought we were going to put the union selector bytes at `a->maxsize` so they don't need to move every time you push or pop from the vector?

quinnj replied:
I started doing it that way, but it seemed to make codgen and mutating operations a lot hairier. I don't quite remember the specifics, but I finally switched over to right after the last element and it seemed overall simpler.

quinnj commented on src/codegen.cpp:
```diff
@@ -6267,7 +6265,7 @@ static void init_julia_llvm_env(Module *m)
     Type *jl_array_llvmt =
         StructType::create(jl_LLVMContext,
-                           makeArrayRef(vaelts),
+                           ArrayRef<Type*>(vaelts,sizeof(vaelts)/sizeof(vaelts[0])),
```
I don't really remember this change; I think it was when we were huddling at JuliaCon. Should I change it back?

vtjnash replied:
It would look simpler, but in the end, makes no difference to the compiler, so not important.

quinnj commented on src/codegen.cpp:
```diff
@@ -2390,18 +2390,35 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
+                    Value *selidx = ctx.builder.CreateMul(emit_arraylen_prim(ctx, ary), nbytes);
+                    selidx = ctx.builder.CreateAdd(selidx, idx);
+                    Value *ptindex = ctx.builder.CreateGEP(T_int8, data, selidx);
+                    Value *tindex = ctx.builder.CreateNUWAdd(ConstantInt::get(T_int8, 1), ctx.builder.CreateLoad(T_int8, ptindex));
```
I took a stab at it using `tbaa_decorate` and `tbaa_const`, maybe it should be `tbaa_immut` though?

vtjnash replied:
It can be a new TBAA node, disjoint from everything else.

quinnj commented:
Update: discovered one reliable crash was due to #22688, involving a Union of the 10 signed/unsigned int types. I was then able to identity some kind of gc error, and have found the minimum example that reliably crashes (on osx at least) to be
```julia
A = Union{Void, Float16, Int8}[]
for i = 1:50
    unshift!(A, Float16(0.0))
end
```
I'd be forever grateful if @yuyichao would take a look or at least give me a few hints of where to look/start debugging. I'm willing to dive into gc stuff, it'll just take me a while to try and grok everything we're doing (I'm coming from about zero there).

vtjnash commented on src/array.c:
```diff
@@ -586,6 +616,10 @@ static int NOINLINE array_resize_buffer(jl_array_t *a, size_t newlen)
         oldnbytes++;
     }
+    if (!a->flags.ptrarray && jl_is_uniontype(jl_tparam0(jl_typeof(a)))) {
+        nbytes += oldlen;
```
`+= newlen`

quinnj replied:
Fixed.

quinnj commented on src/codegen.cpp:
```diff
@@ -2398,18 +2401,35 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
+                    Value *selidx = ctx.builder.CreateMul(emit_arraylen_prim(ctx, ary), nbytes);
+                    selidx = ctx.builder.CreateAdd(selidx, idx);
+                    Value *ptindex = ctx.builder.CreateGEP(T_int8, data, selidx);
+                    Value *tindex = ctx.builder.CreateNUWAdd(ConstantInt::get(T_int8, 1), tbaa_decorate(tbaa_arrayselbyte, ctx.builder.CreateLoad(T_int8, ptindex)));
```
@vtjnash, added the `tbaa_arrayselbyte` tbaa.

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22441#issuecomment-322612034) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2cfe2fe5d694ab3e4754722a613bfdd45d4be82d/8cbbae3_vs_3b51323/report.md). cc @ararslan

quinnj commented:
None of the benchmark results look like anything to worry about; after checking things out locally, it seems like these are just some of the noisiest benchmarks we have.

quinnj commented:
I'm actually not surprised w/ the benchmarks, since I doubt many of them involve Unions at all. What we really need are some isbits-Union benchmarks; I have some lying around, but I'll try to clean them up and post some local results and plan on adding some to BaseBenchmarks.

KristofferC commented:
Tangentially, It would be cool to have some non-micro benchmarks where unions are a natural way of solving the problem, and make the code significantly simpler.

vtjnash commented:
I think the only one I was concerned with was the allocations. Any idea what those came from?

quinnj commented:
Hmmm.....I've spent a few hours now digging thru performance/benchmarking and nothing seems really out of the ordinary. The `problem` benchmarks are pretty tricky to disect since they involve so much random code. I've made a PR to add some basic is bits Unions benchmarks for NanoSoldier [here](https://github.com/JuliaCI/BaseBenchmarks.jl/pull/99).

Anybody else have any ideas on benchmarking or concerns with merging?

quinnj commented:
I'll give this till the morning (12 hours or so) for any last reviewers

StefanKarpinski commented:
![t3relrb](https://user-images.githubusercontent.com/153596/29467176-7fd1d50e-840d-11e7-8c4c-610d27654d92.gif)

tkelman commented:
should have been squashed if the intermediate commits don't all work

quinnj commented:
"all work"; does that mean each commit should pass all tests? just compile? I tried to keep each of the commits pretty logically separated between union field support, array support, then tests, etc.

tkelman commented:
pass tests ideally, depends what future bisectors are looking for.

quinnj commented:
ok, cool. FWIW, I believe tests should have passed for all commits, except maybe after jameson's first, but I also didn't want to clobber his commit w/ the fix.
